packages/block-editor/src/utils/pre-parse-patterns.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, select } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+
+const requestIdleCallback = ( () => {
+	if ( typeof window === 'undefined' ) {
+		return ( callback ) => {
+			setTimeout( () => callback( Date.now() ), 0 );
+		};
+	}
+
+	return window.requestIdleCallback || window.requestAnimationFrame;
+} )();
+
+const cancelIdleCallback = ( () => {
+	if ( typeof window === 'undefined' ) {
+		return clearTimeout;
+	}
+
+	return window.cancelIdleCallback || window.cancelAnimationFrame;
+} )();
+
+export function usePreParsePatterns() {
+	const patterns = useSelect(
+		( _select ) =>
+			_select( blockEditorStore ).getSettings()
+				.__experimentalBlockPatterns,
+		[]
+	);
+
+	useEffect( () => {
+		if ( ! patterns?.length ) {
+			return;
+		}
+
+		let handle;
+		let index = -1;
+
+		const callback = () => {
+			index++;
+			if ( index >= patterns.length ) {
+				return;
+			}
+
+			select( blockEditorStore ).__experimentalGetParsedPattern(
+				patterns[ index ].name,
+				index
+			);
+
+			handle = requestIdleCallback( callback );
+		};
```

Comment:
Have you compared performance between different batching strategies? Roughly:

```js
const patternsPerBatch = 5;

// ...

const callback = () => {
  index += patternsPerBatch;
  if ( index >= patterns.length ) return;

  for (let i = 0; i < patternsPerBatch; i++) {
    select( blockEditorStore ).__experimentalGetParsedPattern(
      patterns[ index + i ].name
    );
  }
```

I suggest trying this out with a few different values for `patternsPerBatch`. In order for the benchmark to mean something, the dummy patterns would have to resemble actual patterns in production — the idea is to determine whether the overhead of setting up `requestIdleCallback` is significant when parsing very small sets of patterns.